pynsot/commands/cmd_circuits.py

```diff
@@ -0,0 +1,378 @@
+# -*- coding: utf-8 -*-
+
+"""
+Sub-command for Circuits
+"""
+
+from __future__ import unicode_literals
+import logging
+
+from nsot.util import slugify
+
+from ..vendor import click
+from . import callbacks
+from .cmd_networks import DISPLAY_FIELDS as NETWORK_DISPLAY_FIELDS
+from .cmd_interfaces import DISPLAY_FIELDS as INTERFACE_DISPLAY_FIELDS
+from .cmd_devices import DISPLAY_FIELDS as DEVICE_DISPLAY_FIELDS
+
+
+# Logger
+log = logging.getLogger(__name__)
+
+# Ordered list of 2-tuples of (field, display_name) used to translate object
+# field names oto their human-readable form when calling .print_list().
+DISPLAY_FIELDS = (
+    ('id', 'ID'),
+    ('name', 'Name'),
+    ('endpoint_a', 'Endpoint A'),
+    ('endpoint_z', 'Endpoint Z'),
+    ('attributes', 'Attributes'),
+)
+
+# Fields to display when viewing a single record.
+VERBOSE_FIELDS = DISPLAY_FIELDS
+
+
+# Main group
+@click.group()
+@click.pass_context
+def cli(ctx):
+    """
+    Circuit objects.
+
+    A Circuit resource represents a connection between two interfaces on a
+    device.
+    """
+
+
+# Add
+# TODO(npegg): support specifying interfaces using the natural key instead of
+# by ID when the API supports that
+@cli.command()
+@click.option(
+    '-a',
+    '--attributes',
+    metavar='ATTRS',
+    help='A key/value pair attached to this Circuit (format: key=value).',
+    multiple=True,
+    callback=callbacks.transform_attributes,
+)
+@click.option(
+    '-A',
+    '--endpoint-a',
+    metavar='INTERFACE_ID',
+    type=int,
+    help='Unique ID of the interface of the A side of the Circuit',
+)
+@click.option(
+    '-n',
+    '--name',
+    metavar='NAME',
+    type=str,
+    help='The name of the Circuit.',
+)
+@click.option(
+    '-s',
+    '--site-id',
+    metavar='SITE_ID',
+    type=int,
+    help='Unique ID of the Site this Circuit is under. [required]',
+    callback=callbacks.process_site_id,
+)
+@click.option(
+    '-Z',
+    '--endpoint-z',
+    metavar='INTERFACE_ID',
+    type=int,
+    help='Unique ID of the interface on the Z side of the Circuit',
+)
+@click.pass_context
+def add(ctx, attributes, endpoint_a, name, site_id, endpoint_z):
+    """
+    Add a new Circuit.
+
+    You must provide the A side of the circuit using the -A/--endpoint-a
+    option. The Z side is recommended but may be left blank, such as in cases
+    where it is not an Interface that is tracked by NSoT (like a provider's
+    interface).
+
+    The name (-n/--name) is optional. If it is not specified, it will be
+    generated for you in the form of:
+    {device_a}:{interface_a}_{device_z}:{interface_z}
+
+    If you wish to add attributes, you may specify the -a/--attributes
+    option once for each key/value pair.
+    """
+
+    data = ctx.params
+
+    if endpoint_a is None:
+        raise click.UsageError('Missing option "-A" / "--endpoint-a"')
```

Comment:
Could we instead mark `-A` as required in the option definition or had you already considered that? If you had, we should alternatively update the help text to include the `[required]` text.